WordPress/gutenberg
# Core CSS support for root padding and alignfull blocks

## What?


Fixes #35607; alternative to #42034.

## Why?


Because root padding doesn't always make sense applied only to the root element (e.g. `body`). As explained in #35607, we should be able to apply it instead to full-width blocks at any level of nesting.

## How?


* Use of this implementation is optional for themes; opt-in via a setting in `theme.json`.
* Attaches custom properties with root padding values to the root element (`body` in the front end, `.editor-styles-wrapper` in the editor); 
* Applies right and left padding to all blocks with content width set (whether default theme widths or custom values)
* Applies negative margins to alignfull children of blocks with root padding set.
* Fixes bug whereby alignfull children of blocks with custom padding wouldn't extend to full width of their container.

UPDATE: As of [02fb1dd](https://github.com/WordPress/gutenberg/pull/42085/commits/02fb1dd7733d657502434ad0745de2782d0204a7), this PR no longer adds the root padding toggle to blocks without content width. This part of the work will be dealt with separately, as suggested [here](https://github.com/WordPress/gutenberg/pull/42085#issuecomment-1178782943)

## Testing Instructions
1. Add `"useRootPaddingAwareAlignments": true` to the settings of your theme's theme.json to enable these changes.
2. Add a Group block to the post and the site editor and set a content width on it.
3. Add an Image block inside the Group and set it to full width.
4. Change root padding value in the site editor by going to Global Styles > Layout > Padding
5. Check that new padding value works in both editors and front end.
6. Check that removing `"useRootPaddingAwareAlignments": true` from theme.json reverts to the previous behaviour (root padding is applied only at the root level, so no blocks extend to full viewport width)

## Screenshots or screencast 

Trunk:

<img width="815" alt="Screen Shot 2022-07-05 at 3 17 39 pm" src="https://user-images.githubusercontent.com/8096000/177254704-24928c63-3d38-4572-8566-f121238e9903.png">

This branch without "useRootVariables" enabled in the theme:

<img width="815" alt="Screen Shot 2022-07-05 at 3 04 36 pm" src="https://user-images.githubusercontent.com/8096000/177254803-d3bb7255-baf2-45bd-a296-06b142ce3e63.png">

This branch with "useRootVariables" enabled in the theme:

<img width="817" alt="Screen Shot 2022-07-05 at 3 04 06 pm" src="https://user-images.githubusercontent.com/8096000/177254886-1558ea94-8069-425c-902b-fc70952533f3.png">
